# HW 4 Coding Q's - Ankit Jain

In [16]:
import numpy as np
import pandas as pd
from statsmodels.formula.api import ols
from plotnine import ggplot, aes, geom_point, labs
from scipy.stats import linregress
import statsmodels.api as sm

In [17]:
df = pd.read_csv("../data/KelleyBlueBookData.csv")
df.head(5)

,Price,Mileage,Make,Model,Trim,Type,Cylinder,Liter,Doors,Cruise,Sound,Leather
0,17314.10313,8221,Buick,Century,Sedan 4D,Sedan,6,3.1,4,1,1,1
1,17542.03608,9135,Buick,Century,Sedan 4D,Sedan,6,3.1,4,1,1,0
2,16218.84786,13196,Buick,Century,Sedan 4D,Sedan,6,3.1,4,1,1,0
3,16336.91314,16342,Buick,Century,Sedan 4D,Sedan,6,3.1,4,1,0,0
4,16339.17032,19832,Buick,Century,Sedan 4D,Sedan,6,3.1,4,1,0,1


In [18]:
# For the dataset KelleyBlueBookData.csv, response= Price against the following predictors: Mileage, Liter, 
# Cylinder (in this order). Treat Cylinder as a quantitative variable.
model = ols('Price ~ Mileage + Liter + Cylinder', data=df).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Price   R-squared:                       0.342
Model:                            OLS   Adj. R-squared:                  0.340
Method:                 Least Squares   F-statistic:                     138.8
Date:                Thu, 25 Sep 2025   Prob (F-statistic):           2.18e-72
Time:                        20:27:55   Log-Likelihood:                -8367.7
No. Observations:                 804   AIC:                         1.674e+04
Df Residuals:                     800   BIC:                         1.676e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   4707.6150   1602.866      2.937      0.003    1561.296    7853.934
Mileage       -0.1544      0.035     -4.461      0.000      -0.222      -0.086
Liter       1545.2522    893.411      1.730      0.084    -208.454    3298.958
Cylinder    2847.9345    712.040      4.000      0.000    1450.247    4245.622
==============================================================================
Omnibus:                      214.158   Durbin-Watson:                   0.074
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              444.825
Skew:                           1.499   Prob(JB):                     2.56e-97
Kurtosis:                       5.071   Cond. No.                     1.37e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.37e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

# 4a. 

In [19]:
sm.stats.anova_lm(model, typ=1) #sequential

,df,sum_sq,mean_sq,F,PR(>F)
Mileage,1.0,1.605590e+09,1.605590e+09,24.890171,7.448191e-07
Liter,1.0,2.421824e+10,2.421824e+10,375.435819,7.134623e-69
Cylinder,1.0,1.031948e+09,1.031948e+09,15.997457,6.930502e-05
Residual,800.0,5.160560e+10,6.450701e+07,NaN,NaN


The null hypothesis is that the coefficient for cylinder is 0, meaning that cylinder is not significant to the model. The alternate hypothesis is that the coefficient is not zero, meaning cylinder is significant. For including the cylinder predictor, the F-stat is 15.997, p-value is <0.001. For excluding the cylinder predictor, the F-stat is 375.436, p-value is <0.001. We reject the null hypothesis because our p-value is <0.001, meaning cylinder is a significant predictor for price.


# 4b. 

In [20]:
y = df['Price']
y_hat = model.predict(df)
mean = df['Price'].mean()
n = len(y)
p = 3

sst = ((y - mean)**2).sum()
sse = ((y - y_hat)**2).sum()

r2 = 1 - (sse/sst)
r2_adjusted = 1 - ((sse / (n - p)) / (sst / (n - 1)))

print(r2)
r2_adjusted

0.34228021178684886


0.3406379651246437

Yes, they match. 

# 4c. 

In [21]:
sm.stats.anova_lm(model, typ=2) #partial

,sum_sq,df,F,PR(>F)
Mileage,1.283997e+09,1.0,19.904763,0.000009
Liter,1.929760e+08,1.0,2.991552,0.084086
Cylinder,1.031948e+09,1.0,15.997457,0.000069
Residual,5.160560e+10,800.0,NaN,NaN


The F-test does not match from 4a because it tests each predictor after factoring in all other predictors instead of sequentially, so the overall result will be different. 

# 4d. 

The null hypothesis is that the coefficient for mileage is 0 after factoring in liter and cylinder, meaning that cylinder is not significant to the model. The alternate hypothesis is that the coefficient is not zero, meaning mileage is significant after factoring in liter and cylinder. For dropping or including the mileage predictor after factoring all other predictors, the F-stat is 19.9, p-value is <0.001. We reject the null hypothesis due to the low p-value. Mileage is a significant predictor to the model after factoring in liter and cylinder. 


# 4e. 

In [43]:
model_null = ols('Price ~ Liter + Cylinder', data=df).fit()
y_hat_null = model_null.predict(df)
SSE_null = ((y - y_hat_null)**2).sum()

F = ((SSE_null - sse) / ((n - 2) - (n-p))) / (sse / (n-p))
F

16.871263806605096

It does not match. 

# 5a. 

In [23]:
model1 = ols('Price ~ Mileage + Cylinder', data=df).fit()
model1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Price   R-squared:                       0.340
Model:                            OLS   Adj. R-squared:                  0.338
Method:                 Least Squares   F-statistic:                     206.2
Date:                Thu, 25 Sep 2025   Prob (F-statistic):           5.95e-73
Time:                        20:27:55   Log-Likelihood:                -8369.2
No. Observations:                 804   AIC:                         1.674e+04
Df Residuals:                     801   BIC:                         1.676e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   3145.7503   1325.934      2.372      0.018     543.034    5748.467
Mileage       -0.1524      0.035     -4.401      0.000      -0.220      -0.084
Cylinder    4027.6746    204.612     19.684      0.000    3626.036    4429.313
==============================================================================
Omnibus:                      198.944   Durbin-Watson:                   0.077
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              385.493
Skew:                           1.439   Prob(JB):                     1.96e-84
Kurtosis:                       4.797   Cond. No.                     1.01e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.01e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

R^2 is 0.34 and R^2 adjusted is 0.338

# 5b. 

In [24]:
y = df['Price']
y_hat = model1.predict(df)
mean = df['Price'].mean()
n = len(y)
p = 3

sst = ((y - mean)**2).sum()
sse = ((y - y_hat)**2).sum()

r2 = 1 - (sse/sst)
r2_adjusted = 1 - ((sse / (n - p)) / (sst / (n - 1)))

print(r2)
r2_adjusted

0.33982070826263056


0.338172320517968

Yes, they match.

# 5c. 

In [25]:
model2 = ols('Price ~ Mileage + Liter + Cylinder', data=df).fit()
model2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Price   R-squared:                       0.342
Model:                            OLS   Adj. R-squared:                  0.340
Method:                 Least Squares   F-statistic:                     138.8
Date:                Thu, 25 Sep 2025   Prob (F-statistic):           2.18e-72
Time:                        20:27:55   Log-Likelihood:                -8367.7
No. Observations:                 804   AIC:                         1.674e+04
Df Residuals:                     800   BIC:                         1.676e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   4707.6150   1602.866      2.937      0.003    1561.296    7853.934
Mileage       -0.1544      0.035     -4.461      0.000      -0.222      -0.086
Liter       1545.2522    893.411      1.730      0.084    -208.454    3298.958
Cylinder    2847.9345    712.040      4.000      0.000    1450.247    4245.622
==============================================================================
Omnibus:                      214.158   Durbin-Watson:                   0.074
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              444.825
Skew:                           1.499   Prob(JB):                     2.56e-97
Kurtosis:                       5.071   Cond. No.                     1.37e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.37e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

The R^2 value is 0.342 and adjusted R^2 is 0.34. According to adjusted R^2, model 2 is preferable because it has a higher adjusted R^2. This is because adjusted R^2 adjusts for the fact that R^2 will always increase when more predictors are added, meaning that the base R^2 is not necesarilly the best predictor. 

# 5d. 

In [44]:
sm.stats.anova_lm(model1, typ=2) #partial

,sum_sq,df,F,PR(>F)
Mileage,1.252375e+09,1.0,19.366403,1.224901e-05
Cylinder,2.505721e+10,1.0,387.478325,1.156299e-70
Residual,5.179858e+10,801.0,NaN,NaN


In [45]:
sm.stats.anova_lm(model1, typ=1) #sequential

,df,sum_sq,mean_sq,F,PR(>F)
Mileage,1.0,1.605590e+09,1.605590e+09,24.828439,7.681214e-07
Cylinder,1.0,2.505721e+10,2.505721e+10,387.478325,1.156299e-70
Residual,801.0,5.179858e+10,6.466739e+07,NaN,NaN


In [46]:
sm.stats.anova_lm(model2, typ=2) #partial

,sum_sq,df,F,PR(>F)
Mileage,1.283997e+09,1.0,19.904763,0.000009
Liter,1.929760e+08,1.0,2.991552,0.084086
Cylinder,1.031948e+09,1.0,15.997457,0.000069
Residual,5.160560e+10,800.0,NaN,NaN


In [47]:
sm.stats.anova_lm(model2, typ=1) #sequential

,df,sum_sq,mean_sq,F,PR(>F)
Mileage,1.0,1.605590e+09,1.605590e+09,24.890171,7.448191e-07
Liter,1.0,2.421824e+10,2.421824e+10,375.435819,7.134623e-69
Cylinder,1.0,1.031948e+09,1.031948e+09,15.997457,6.930502e-05
Residual,800.0,5.160560e+10,6.450701e+07,NaN,NaN


I would say the best model is Model 1 because it has tested better with ANOVA and the t-tests. The R^2 adjusted is slightly lower on Model 1 but it is neglible when you see that the t-test for the liter coefficient has a p-value of 0.08 which makes it a weaker predictor. ANOVA also shows liter being a wekaer predictor. Since liter is not adding significance to my model, I prefer going with model 1.